File: 08_Triton/02_softmax.c
#include <stdio.h>
#include <math.h>

/*
===========================================
Softmax 函数的 C 语言实现（含数值稳定性优化）

功能：
- 对一个长度为 n 的 float 数组 x[] 进行 softmax 归一化处理
- 包含数值稳定性处理（减去最大值 max），避免浮点溢出
- 示例输入：x = [1.0, 2.0, 3.0]
- 经过 softmax 后输出约为：[0.0900, 0.2447, 0.6652]

softmax 公式：
    softmax(x_i) = exp(x_i) / sum(exp(x_j))

数值稳定性技巧：
    为防止溢出，先减去 max(x)，即：
    softmax(x_i) = exp(x_i - max) / sum(exp(x_j - max))
===========================================
*/

void softmax(float *x, int n) {
    float max = x[0];

    // 第一步：找到数组中的最大值（用于数值稳定）
    for (int i = 1; i < n; i++) {
        if (x[i] > max) {
            max = x[i];
        }
    }

    float sum = 0.0;

    // 第二步：将每个元素减去最大值后进行 exp 计算
    for (int i = 0; i < n; i++) {
        x[i] = exp(x[i] - max);  // 减 max 是为了防止 exp 溢出
        sum += x[i];             // 同时累加求和
    }

    // 第三步：将每个值除以总和，得到最终的概率值
    for (int i = 0; i < n; i++) {
        x[i] /= sum;
    }
}

int main() {
    // 示例输入数组
    float x[] = {1.0, 2.0, 3.0};

    // 调用 softmax 函数
    softmax(x, 3);

    // 输出结果
    for (int i = 0; i < 3; i++) {
        printf("%f\n", x[i]);
    }

    return 0;
}

File: 08_Triton/02_softmax.py
"""
===========================================================
Triton × PyTorch Softmax 对比示例
-----------------------------------------------------------
功能概览
1. 使用 **Triton** 编写并 JIT 编译一个按行 Softmax 的 GPU Kernel。
2. 调用 `triton_softmax` 计算 2‑D Tensor（形状：n_rows × n_cols）的
   softmax，并与 **PyTorch** 的 `torch.softmax` 结果对比。
3. 打印两者最大绝对误差与一致性判断。

核心要点
▶ 数值稳定性：先减去每行最大值再做 exp，避免溢出。  
▶ `BLOCK_SIZE` 取列数的最近 2 次幂，且不超过 1024。  
▶ Kernel 通过 `tl.arange` + mask 只加载行内有效元素。  
===========================================================
"""

import torch
import triton
import triton.language as tl

# --------------------------------------------------------
# Triton Kernel：按行 Softmax（数值稳定版）
# --------------------------------------------------------
@triton.jit
def softmax_kernel(
    output_ptr,             # ➜ 输出张量首地址
    input_ptr,              # ➜ 输入张量首地址
    input_row_stride,       # ➜ 输入张量行步长
    output_row_stride,      # ➜ 输出张量行步长
    n_cols,                 # ➜ 每行元素数
    BLOCK_SIZE: tl.constexpr,  # ➜ 每行一次加载的 BLOCK 大小（2 次幂）
):
    # 当前 program 对应的行号 (row_idx)
    row_idx = tl.program_id(axis=0)

    # 计算当前行在内存中的起始地址
    row_start_ptr    = input_ptr  + row_idx * input_row_stride
    out_row_start_ptr = output_ptr + row_idx * output_row_stride

    # -------- ① 载入一行数据到 SRAM --------
    cols = tl.arange(0, BLOCK_SIZE)                       # 0..BLOCK_SIZE-1
    mask = cols < n_cols                                  # 防止越界
    row  = tl.load(row_start_ptr + cols, mask=mask,
                   other=-float('inf'))                   # 越界填 -inf

    # -------- ② 数值稳定：减 max 后做 exp --------
    row_max   = tl.max(row, axis=0)                       # 每行最大值
    numerator = tl.exp(row - row_max)                     # e^(x - max)

    # -------- ③ 求和并归一化 --------
    denominator     = tl.sum(numerator, axis=0)           # ∑ e^(x - max)
    softmax_output  = numerator / denominator             # softmax

    # -------- ④ 写回结果 --------
    tl.store(out_row_start_ptr + cols, softmax_output, mask=mask)


# --------------------------------------------------------
# Triton 包装函数：自动确定 BLOCK_SIZE 并调用 Kernel
# --------------------------------------------------------
def triton_softmax(x: torch.Tensor):
    n_rows, n_cols = x.shape
    output = torch.empty_like(x)

    # BLOCK_SIZE 取列数的最近 2 次幂，最大 1024
    BLOCK_SIZE = triton.next_power_of_2(n_cols)
    BLOCK_SIZE = min(BLOCK_SIZE, 1024)

    # Launch grid：每个 program 处理一行
    grid = (n_rows,)

    softmax_kernel[grid](
        output, x,
        x.stride(0), output.stride(0),  # 行步长（元素数）
        n_cols,
        BLOCK_SIZE=BLOCK_SIZE
    )
    return output


# ===================== 测试 =====================
torch.manual_seed(0)
x = torch.randn(256, 1024, device='cuda')  # 256×1024 随机输入

# PyTorch 参考结果
torch_result = torch.softmax(x, dim=1)

# Triton 结果
triton_result = triton_softmax(x)

# 误差统计
max_diff = torch.max(torch.abs(torch_result - triton_result))
print(f"Maximum difference between PyTorch and Triton results: {max_diff:.2e}")

# 一致性判断
is_close = torch.allclose(torch_result, triton_result, rtol=1e-5, atol=1e-5)
print(f"Results are close: {is_close}")
